src/Processors/Sources/SourceWithProgress.cpp

```diff
@@ -23,6 +23,13 @@ SourceWithProgress::SourceWithProgress(Block header, bool enable_auto_progress)
 {
 }
 
+void SourceWithProgress::setProcessListElement(QueryStatus * elem)
+{
+    process_list_elem = elem;
+    /// Update total_rows_approx as soon as possible.
+    progress({});
```

Comment:
Maybe do only this part of progress?
https://github.com/ClickHouse/ClickHouse/blob/ddca262fe63445d9ea077b773926692b25356880/src/Processors/Sources/SourceWithProgress.cpp#L49-L60

Reply:
>Maybe do only this part of progress?

Yep, good idea.